docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+
+```csharp
+struct ErrorData
+{
+    public int Code;
+    public bool IsFatalError;
+    public string? Message;
+}
+```
+
+Following the naming pattern, we will name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, we will only define marshallers for some modes. This will demonstrate the case where if a mode isn't provided, the source generator will fail. Let's start with the "in" direction.
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
+    internal static unsafe class ErrorDataMarshaller
+    {
+        // Unmanaged representation of ErrorData.
+        // Should mimic the unmanaged error_data type at a binary level.
+        internal struct ErrorDataUnmanaged
+        {
+            public int Code;        // .NET doesn't support less than 32-bit, so int is 32-bit.
+            public byte IsFatal;    // The C++ bool is defined as a single byte.
+            public uint* Message;   // This could be as simple as a void*, but uint* is closer.
+        }
+
+        public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+            => throw new NotImplementedException();
+
+        public static void Free(ErrorDataUnmanaged unmanaged)
+            => throw new NotImplementedException();
+    }
+}
+```
+
+The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, that will mimic the shape of the unmanaged type is now trivial with `Utf32StringMarshaller`. Note this is a "stateless" marshaller because the marshaller itself only consists of `static` functions.
+
+Marshalling of an `int` is unnecessary since representations are identical in unmanaged and managed. A `bool`'s binary representation isn't defined in .NET so we use its current value to define a zero and non-zero value in the unmanaged type. We then reuse our UTF-32 marshaller to convert our `string` field into a `uint*`.
+
+```csharp
+public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+{
+    return new ErrorDataUnmanaged
+    {
+        Code = managed.Code,
+        IsFatal = (byte)(managed.IsFatalError ? 1 : 0),
+        Message = Utf32StringMarshaller.ConvertToUnmanaged(managed.Message),
+    };
+}
+```
+
+Recall we are defining this marshaller as an "in" so we must clean-up any allocations performed during the marshalling. The `int` and `bool` fields didn't allocate any memory, but the `Message` field did. We again reuse `Utf32StringMarshaller` to clean up the marshalled string.
+
+```csharp
+public static void Free(ErrorDataUnmanaged unmanaged)
+    => Utf32StringMarshaller.Free(unmanaged.Message);
+```
+
+Next, we are going to consider the case where multiple instances of `error_data` are returned. Typically returning multiple of a thing is done using a collection and those multiple things are called "elements". Since this will be returning multiple elements we call this the `MarshalMode.ElementOut` mode.
```

Comment:
```suggestion
Next, consider the case where multiple instances of `error_data` are returned. Typically, you use a collection to return multiple elements. Since this marshaller will be returning multiple elements, this is called the `MarshalMode.ElementOut` mode.
```